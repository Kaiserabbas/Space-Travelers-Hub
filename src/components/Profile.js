import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { leaveMission } from '../redux/missions/missionsSlice';
import '../assets/css/profile.css';

function Profile() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  const joinedMissions = missions.filter((mission) => mission.reserved);

  const rockets = useSelector((state) => state.rockets.rockets);
  const reservedRockets = rockets.filter((rocket) => rocket.reserved);

  const [showCancelId, setShowCancelId] = useState(null);

  const handleClick = (id) => {
    if (showCancelId === id) {
      setShowCancelId(null);
    } else {
      setShowCancelId(id);
    }
  };

  const handleLeaveMission = (id) => {
    dispatch(leaveMission(id));
  };
  return (
    <div>
      <div className="profile">
        <div>
          <h2>Rockets</h2>
          <ul className="rockets-profile">
            {reservedRockets.map((rocket) => (
              <li className="rockets-list" key={rocket.id}>
                <span className="list-item">
                  {rocket.rocket_name}
                </span>
              </li>
            ))}
          </ul>
        </div>
        <div>
          <div>
            <h2>Missions</h2>
            <ul className="missions-profile">
              {joinedMissions.map((mission) => (
                <li key={mission.mission_id} className="missions-list">
                  <span
                    className="mission-name"
                    onClick={() => handleClick(mission.mission_id)}
                    role="button"
                    tabIndex={0}
                    onKeyDown={(e) => {
                      if (e.key === 'Enter') {
                        handleClick();
                      }
                    }}
                  >
                    {mission.mission_name}
                    {showCancelId === mission.mission_id && (
                    <button
                      className="missions-cancel"
                      type="button"
                      onClick={(e) => {
                        e.stopPropagation();
                        handleLeaveMission(mission.mission_id);
                      }}
                    >
                      Cancel Reservation
                    </button>
                    )}
                  </span>
                </li>
              ))}
            </ul>
          </div>
        </div>
      </div>
    </div>
  );
}

export default Profile;
