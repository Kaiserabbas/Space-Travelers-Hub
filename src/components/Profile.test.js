import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Profile from './Profile';
import '@testing-library/jest-dom/extend-expect';

// Setting up mock store
const mockStore = configureMockStore();

// Mock state data
const mockState = {
  missions: {
    missions: [
      {
        mission_id: '1',
        mission_name: 'Sample Mission',
        reserved: true,
      },
    ],
  },
  rockets: {
    rockets: [
      {
        id: '1',
        rocket_name: 'Sample Rocket',
        reserved: true,
      },
    ],
  },
};

describe('Profile Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore(mockState);
  });

  // Snapshot Test
  test('Profile matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Profile />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
