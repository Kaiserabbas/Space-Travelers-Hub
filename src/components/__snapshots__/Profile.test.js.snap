// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component Profile matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="profile"
  >
    <div
      class="missions-profile"
    >
      <h2>
        My Missions
      </h2>
      <ul>
        <li
          class="missions-list"
        >
          <span
            class="mission-name"
            role="button"
            tabindex="0"
          >
            Sample Mission
          </span>
        </li>
      </ul>
    </div>
    <div
      class="rockets-profile"
    >
      <h2>
        My Rockets
      </h2>
      <ul>
        <li
          class="rockets-list"
        >
          <span
            class="rocket-name"
          >
            Sample Rocket
          </span>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
